pkg/osquery/tables/systemprofiler/systemprofiler.go

```diff
@@ -0,0 +1,236 @@
+	"context"
+	"os/exec"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/groob/plist"
+	"github.com/kolide/launcher/pkg/dataflatten"
+	"github.com/kolide/osquery-go"
+	"github.com/kolide/osquery-go/plugin/table"
+	"github.com/pkg/errors"
+)
+
+const systemprofilerPath = "/usr/sbin/system_profiler"
+
+var knownDetailLevels = []string{
+	"mini",  // short report (contains no identifying or personal information)
+	"basic", // basic hardware and network information
+	"full",  // all available information
+}
+
+type Property struct {
+	Order                string `plist:"_order"`
+	SuppressLocalization string `plist:"_suppressLocalization"`
+	DetailLevel          string `plist:"_detailLevel"`
+}
+
+type Result struct {
+	Items          []interface{} `plist:"_items"`
+	DataType       string        `plist:"_dataType"`
+	SPCommandLine  []string      `plist:"_SPCommandLineArguments"`
+	ParentDataType string        `plist:"_parentDataType"`
+
+	// These would be nice to add, but they come back with inconsistent
+	// types, so doing a straight unmarshal is hard.
+	// DetailLevel    int                 `plist:"_detailLevel"`
+	// Properties     map[string]Property `plist:"_properties"`
+}
+
+type Table struct {
+	client    *osquery.ExtensionManagerClient
+	logger    log.Logger
+	tableName string
+}
+
+func TablePlugin(client *osquery.ExtensionManagerClient, logger log.Logger) *table.Plugin {
+
+	columns := []table.ColumnDefinition{
+		table.TextColumn("fullkey"),
+		table.TextColumn("parent"),
+		table.TextColumn("key"),
+		table.TextColumn("value"),
+		table.TextColumn("parentdatatype"),
+
+		table.TextColumn("query"),
+		table.TextColumn("datatype"),
+		table.TextColumn("detaillevel"),
+	}
+
+	t := &Table{
+		client:    client,
+		logger:    level.NewFilter(logger, level.AllowInfo()),
+		tableName: "kolide_system_profiler",
+	}
+
+	return table.NewPlugin(t.tableName, columns, t.generate)
+}
+
+func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	var results []map[string]string
+
+	requestedDatatypes := []string{}
+
+	datatypeQ, ok := queryContext.Constraints["datatype"]
+	if !ok || len(datatypeQ.Constraints) == 0 {
+		return results, errors.Errorf("The %s table requires that you specify a constraint for datatype", t.tableName)
+	}
+
+	for _, datatypeConstraint := range datatypeQ.Constraints {
+		dt := datatypeConstraint.Expression
+
+		// If the constraint is the magic "%", it's eqivlent to an `all` style
+		if dt == "%" {
+			requestedDatatypes = []string{}
+			break
+		}
+
+		requestedDatatypes = append(requestedDatatypes, dt)
+	}
+
+	var detailLevel string
+	if q, ok := queryContext.Constraints["detaillevel"]; ok && len(q.Constraints) != 0 {
+		if len(q.Constraints) > 1 {
+			level.Info(t.logger).Log("msg", "WARNING: Only using the first detaillevel request")
+		}
+
+		dl := q.Constraints[0].Expression
+		for _, known := range knownDetailLevels {
```

Comment:
why make `knownDetailLevels` an array instead of a set(map)?

Reply:
AFAIK this is the preferred go idiom. 

If I needed to do something with the resultant strings (like map them to numeric values) I think a `map` would be cleaner, but for checking known values this is super common